lighthouse-core/audits/image-size-responsive.js

```diff
@@ -213,12 +213,25 @@ class ImageSizeResponsive extends Audit {
    */
   static audit(artifacts) {
     const DPR = artifacts.ViewportDimensions.devicePixelRatio;
-    const results = Array
+
+    const validImages = Array
       .from(artifacts.ImageElements)
       .filter(isCandidate)
-      .filter(image => !imageHasRightSize(image, DPR))
-      .filter(image => isVisible(image.clientRect, artifacts.ViewportDimensions))
-      .map(image => getResult(image, DPR));
+      .filter(image => {
+        const {naturalWidth, naturalHeight} = image;
```

Comment:
you can inline this check into `imageHasRightSize` and change its return type to  `image is LH.Artifacts.ImageElement & {naturalWidth: number, naturalHeight: number}` to clean up most of this :)

see [ts playground example](https://www.typescriptlang.org/play?ts=3.9.7&useJavaScript=true#code/PQKhAIAEBcE8AcCmATRAzcBvTAPAXOAHYCuAtgEaIBOANOLAPwEkXUC+b4AYgPY-ghgAKCGgQQgVHgBDKtNJZePTmj4SIkKomjEqhLKv4BLAM7c+4AGRZYzMpSod1wtMUIBjaEZ76A5jwAKQwBKLAlwCK0dPXA4JB4MQwA6WHAAXgzwAHIWByyhNhEhdx8TaHBDdKx8cABGQqMMAP8gvmDQksITHgAbRCSenl9WnhTgoA)

Reply:
@patrickhulce `imageHasRightSize` is doing more than a type check, and it's used to enforce the inverse of the result. I think it makes sense to at least switch the return type back to `boolean`.

-----
Additionally...

We could split the type check into another function like `imageHasNaturalDimensions` and use it in an additional filter on L221.

Example:
```js
/**
 * @param {LH.Artifacts.ImageElement} image 
 * @return {image is LH.Artifacts.ImageElement & {naturalWidth: number, naturalHeight: number}}
 */
function imageHasNaturalDimensions(image) {
 return image.naturalHeight !== undefined && image.naturalWidth !== undefined;
}

/**
 * @param {LH.Artifacts.ImageElement & {naturalWidth: number, naturalHeight: number}} image
 * @param {number} DPR
 * @return {boolean}
 */
function imageHasRightSize(image, DPR) {
  const [expectedWidth, expectedHeight] =
      allowedImageSize(image.displayedWidth, image.displayedHeight, DPR);
  return image.naturalWidth >= expectedWidth && image.naturalHeight >= expectedHeight;
}
```
And the filter:
```js
    const results = Array
      .from(artifacts.ImageElements)
      .filter(isCandidate)
      .filter(imageHasNaturalDimensions)
      .filter(image => !imageHasRightSize(image, DPR))
      .filter(image => isVisible(image.clientRect, artifacts.ViewportDimensions))
      .map(image => getResult(image, DPR));
```

Reply:
splitting into 2 sounds good to me 👍

Reply:
@patrickhulce Does this still bring up the ts issue we had? From what I remember, when it reaches the `!imageHasRightSize` call, ts will complain again that the `image`'s natural dimensions could be `undefined`.

Reply:
Yes adam's example here should still fix that situation we ran into, feel free to ping me if it's still giving you trouble and we can walk through 👍